include/boost/random/detail/qrng_base.hpp

```diff
@@ -0,0 +1,203 @@
+  }
+
+  //!Writes a @c DerivedT to a @c std::ostream.
+  BOOST_RANDOM_DETAIL_OSTREAM_OPERATOR(os, DerivedT, s)
+  {
+    os << s.dimension() << " " << s.seq_count << " " << s.curr_elem;
+    return os;
+  }
+
+  //!Reads a @c DerivedT from a @c std::istream.
+  BOOST_RANDOM_DETAIL_ISTREAM_OPERATOR(is, DerivedT, s)
+  {
+    std::size_t dim, seed, z;
+    if (is >> dim >> std::ws >> seed >> std::ws >> z) // initialize iff success!
+    {
+      if (s.dimension() != dim)
+      {
+        prevent_zero_dimension(dim);
+        s.lattice.resize(dim);
+        s.quasi_state.resize(dim);
+      }
+      // Fast-forward to the correct state
+      s.seed(seed);
+      s.discard(z);
+    }
+    return is;
+  }
+
+  //!Returns true if the two generators will produce identical sequences.
+  BOOST_RANDOM_DETAIL_EQUALITY_OPERATOR(DerivedT, x, y)
+  {
+    const std::size_t dimension_value = x.dimension();
+
+    // Note that two generators with different seq_counts and curr_elems can
+    // produce the same sequence because the generator triple
+    // (D, S, D) is equivalent to (D, S + 1, 0), where D is dimension, S -- seq_count,
+    // and the last one is curr_elem.
+
+    return (dimension_value == y.dimension()) &&
+      (x.seq_count + (x.curr_elem / dimension_value) == y.seq_count + (y.curr_elem / dimension_value)) &&
+      (x.curr_elem % dimension_value == y.curr_elem % dimension_value);
+  }
+
+  //!Returns true if the two generators will produce different sequences,
+  BOOST_RANDOM_DETAIL_INEQUALITY_OPERATOR(DerivedT)
+
+protected:
+  DerivedT& derived() throw()
+  {
+    return *static_cast<DerivedT * const>(this);
+  }
+
+  void reset_state()
+  {
+    curr_elem = 0;
+    seq_count = 0;
+    std::fill(quasi_state.begin(), quasi_state.end(), result_type /*zero*/());
+  }
+
+private:
+  inline static std::size_t prevent_zero_dimension(std::size_t dimension)
+  {
+    if (dimension == 0)
+      boost::throw_exception( std::invalid_argument("qrng_base: zero dimension") );
+    return dimension;
+  }
+
+  // Load the result from the saved state.
+  result_type load_cached()
+  {
+    return quasi_state[curr_elem++];
+  }
+
+  result_type next_state()
+  {
+    derived().compute_next();
+
+    curr_elem = 0;
+    return load_cached();
+  }
+
+  // Discards z consecutive s-dimensional vectors,
+  // and preserves the position of the element-to-read
+  void discard_vector(std::size_t z)
+  {
+    std::size_t inc_seq_count = seq_count + z;
+    // Here we check that no overflow occurs before we
+    // begin seeding the new value
+    if (inc_seq_count > seq_count)
+    {
+      std::size_t tmp = curr_elem;
+
+      derived().seed(inc_seq_count);
+
+      curr_elem = tmp;
+    }
+    else if (inc_seq_count < seq_count) // Increment overflowed?
+    {
+      boost::throw_exception( std::overflow_error("discard_vector") );
+    }
+  }
+
+protected:
+  LatticeT lattice;
+  std::size_t curr_elem;
+  std::size_t seq_count;
```

Comment:
In my Atanassov and Halton sequence, I wanted to be able to consume the stream from a multithreaded context, so I made the sequence count atomic. Am I correct in understanding that (1) currently the stream cannot be consumed from a multithreaded context without a race condition, and (2) that making `curr_elem` and `seq_count` atomic would allow consumption from a multithreaded context?

Reply:
I need to think about this, but at this particular moment am pretty averse to anything that makes the other users pay for what they don't use. So I wouldn't want to just slap on std::atomic here. What do you think, @swatanabe ?